docker-build

```diff
@@ -64,12 +103,49 @@ push() {
 
 run() {
   local repo="$1"
-  build "$repo" "$(pull "$repo")" && push "$repo"
+  shift
+  build_with_cache "$repo" "$@" && push "$repo"
 }
 
+# Main -------------------------------------------------
+
 case "$1" in
+  help|-h|--help) usage ;;
+esac
+
+if [ -n "$CIRCLECI" ]; then
+  REPO="$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME"
+  BRANCH="$CIRCLE_BRANCH"
+  SHA="$CIRCLE_SHA1"
+else
+  >&2 echo "This script only works on CircleCI right now."
+  exit 1
+fi
+
+if [ "$DOCKER_BUILDKIT" = 1 ]; then
+  docker_version="$(docker version -f "{{.Server.Version}}")"
+  docker_version_maj="$(echo "$docker_version" | cut -d'.' -f 1)"
+  docker_version_min="$(echo "$docker_version" | cut -d'.' -f 2)"
+  if [ "$docker_version_maj" -lt 19 ] \
+    || { [ "$docker_version_maj" -eq 19 ] && [ "$docker_version_min" -lt 3 ]; }; then
+    >&2 echo "Docker engine version ${docker_version} is too old, must be >= 19.03 to use buildkit"
+    exit 1
+  fi
+fi
+
+if [ -n "$NO_EXTRA_TAGS" ]
+then
+    extra_tags=false
+else
+    extra_tags=true
+fi
+
+
+cmd="$1"
+case "$cmd" in
   pull) pull "$REPO" ;;
-  build) build "$REPO" ;;
+  build) shift; build "$REPO" "$@" ;;
+  build-with-cache) shift; build_with_cache "$REPO" "$@" ;;
   push) push "$REPO" ;;
-  *) run "$REPO" ;;
+  *) run "$REPO" "$@" ;;
```

Comment:
Any reason not to parse out the `cmd` before checking for the help case? I find it more confusing that we have 2 separate case statements.

Reply:
I wanted to handle the help/usage command before the check for `$CIRCLECI` so that at least that part could be run outside circle. The regular command case switch has to happen after that section though so that globals like `$REPO` are set.